File: [K]graph/[K]graph-bfs/200-number-of-islands.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols = len(grid), len(grid[0])
        visited = set()
        res = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r, c) not in visited:
                    res += 1
                    visited.add((r, c))
                    queue = deque([(r, c)])
                    while queue:
                        row, col = queue.popleft()
                        for next_r, next_c in [(row+1, col), (row-1, col), (row, col+1), (row, col-1)]:
                            if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited or grid[next_r][next_c] == "0":
                                continue
                            queue.append((next_r, next_c))
                            visited.add((next_r, next_c))
        return res
    
# time O(RC)
# space O(RC)
# using graph and bfs with single source

File: [K]graph/[K]graph-bfs/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int res = 0;
        int rows = grid.length;
        int cols = grid[0].length;
        HashSet<Integer> visited = new HashSet<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (visited.contains(r * cols + c) || grid[r][c] == '0') {
                    continue;
                }
                ArrayDeque<int[]> queue = new ArrayDeque<>();
                queue.offer(new int[]{r, c});
                visited.add(r * cols + c);
                res += 1;
                while (!queue.isEmpty()) {
                    int[] rc = queue.poll();
                    int cur_r = rc[0];
                    int cur_c = rc[1];
                    for (int[] nextrc: new int[][]{{cur_r+1, cur_c}, {cur_r-1, cur_c}, {cur_r, cur_c+1}, {cur_r, cur_c-1}}) {
                        int next_r = nextrc[0];
                        int next_c = nextrc[1];
                        if (0 <= next_r && next_r < rows && 0 <= next_c && next_c < cols && !visited.contains(next_r * cols + next_c) && grid[next_r][next_c] == '1') {
                            queue.offer(new int[]{next_r, next_c});
                            visited.add(next_r * cols + next_c);
                        }
                    }
                }
            }
        }
        return res;
    }
}

// time O(RC)
// space O(RC)
// using graph and bfs with single source